File: android/src/main/kotlin/com/podoc/flutter_youtube_player/PlayerViewLease.kt
package com.podoc.flutter_youtube_player

import android.app.Activity
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView

internal class PlayerViewLease(
  private val pool: PlayerViewPool,
  private val playerView: NativePlayerView,
  messenger: BinaryMessenger,
  viewId: Int,
  hostActivity: Activity?,
) : PlatformView {
  private var disposed = false

  init {
    playerView.bind(messenger, viewId, hostActivity)
  }

  override fun getView(): View = playerView.rootView

  override fun dispose() {
    if (disposed) return
    disposed = true
    playerView.unbind()
    pool.releasePlayerView(playerView)
  }
}

File: ios/Classes/PlayerViewLease.swift
import Flutter
import UIKit

final class PlayerViewLease: NSObject, FlutterPlatformView {
  private let pool: PlayerViewPool
  private let playerView: NativePlayerView

  init(
    pool: PlayerViewPool,
    playerView: NativePlayerView,
    viewId: Int64,
    messenger: FlutterBinaryMessenger
  ) {
    self.pool = pool
    self.playerView = playerView
    super.init()
    playerView.bind(
      viewId: viewId,
      messenger: messenger
    )
  }

  func view() -> UIView { playerView.rootView }

  deinit {
    playerView.unbind()
    pool.release(playerView)
  }
}
